xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,399 @@
+		gsb.cc.UpdateState(state)
+		return
+	}
+	// If the current balancer is currently in a state other than READY, the
+	// new policy can be swapped into place immediately. This is because
+	// there is no reason to gracefully switch from and keep using the old
+	// policy as the ClientConn is not connected to any backends.
+	if state.ConnectivityState != connectivity.Connecting || gsb.balancerCurrent.lastState.ConnectivityState != connectivity.Ready {
+		gsb.swap()
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *Balancer) swap() {
+	gsb.cc.UpdateState(gsb.balancerPending.lastState)
+	cur := gsb.balancerCurrent
+	gsb.balancerCurrent = gsb.balancerPending
+	gsb.balancerPending = nil
+	go func() {
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		cur.Close()
+	}()
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *Balancer) balancerCurrentOrPending(bw *balancerWrapper) bool {
+	return bw == gsb.balancerCurrent || bw == gsb.balancerPending
+}
+
+// caller must hold gsb.mu.
+func (gsb *Balancer) newSubConn(bw *balancerWrapper, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if !gsb.balancerCurrentOrPending(bw) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bw, bw)
+	}
+	return gsb.cc.NewSubConn(addrs, opts)
+}
+
+func (gsb *Balancer) resolveNow(bw *balancerWrapper, opts resolver.ResolveNowOptions) {
+	// Ignore ResolveNow requests from anything other than the most recent
+	// balancer, because older balancers were already removed from the config.
+	if bw == gsb.latestBalancer() {
+		gsb.cc.ResolveNow(opts)
+	}
+}
+
+// caller must hold gsb.mu.
+func (gsb *Balancer) removeSubConn(bw *balancerWrapper, sc balancer.SubConn) {
+	if !gsb.balancerCurrentOrPending(bw) {
+		return
+	}
+	gsb.cc.RemoveSubConn(sc)
+}
+
+func (gsb *Balancer) updateAddresses(bw *balancerWrapper, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bw) {
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *Balancer) SwitchTo(builder balancer.Builder) error {
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return errBalancerClosed
+	}
+	bw := &balancerWrapper{
+		gsb:       gsb,
+		lastState: defaultState,
+		subconns:  make(map[balancer.SubConn]bool),
+	}
+	var balToClose balancer.Balancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = bw
+	} else {
+		// Clean up resources here that are from a previous pending lb.
+		if gsb.balancerPending != nil {
+			for sc := range gsb.balancerPending.subconns {
+				gsb.cc.RemoveSubConn(sc)
+			}
```

Comment:
Delete?

Then simplify the whole thing to just `else { balToClose = gsb.balancerPending; gsb.balancerPending=bw }`

Reply:
Wait, actually, I had to keep the balToClose part inside the nil check. I remember the problem. balancerPending being nil (0x0, 0x0) for both parts of the interface, when I would do newBalancer = balancerPending unconditionally, newBalancer would not be (0x0, 0x0) for both parts of the interface ever, even when gsb.balancerPending was.

Reply:
Switched to balToClose declared as *balancerWrapper, then your change worked.